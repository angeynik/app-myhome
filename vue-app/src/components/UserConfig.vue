<template>
    <div>
      <h1>Управление пользователями</h1>
  
      <!-- Форма для создания пользователя -->
      <div>
        <h2>Создать нового пользователя</h2>
        <form @submit.prevent="createUser">
          <div>
            <label for="username">Имя пользователя:</label>
            <input type="text" id="username" v-model="newUser.username" required />
          </div>
          <p></p>
          <div>
            <label for="password">Пароль:</label>
            <input type="password" id="password" v-model="newUser.password" required />
          </div>
          <p></p>
          <div>
            <label for="dataSource">Источник данных:</label>
            <input type="text" id="dataSource" v-model="newUser.dataSource" required />
          </div>
          <p></p>
          <div>
            <label for="permissionLevel">Уровень доступа:</label>
            <select id="permissionLevel" v-model="newUser.permissionLevel" required>
              <option value="1">Чтение</option>
              <option value="2">Запись</option>
              <option value="3">Администрирование</option>
            </select>
          </div>
          <p></p>
          <button type="submit">Создать пользователя</button>
        </form>
      </div>
  
      <!-- Сообщение об успешном создании пользователя -->
      <div v-if="successMessage" class="success-message">
        {{ successMessage }}
      </div>
  
      <!-- Список существующих пользователей -->
      <div>
        <h2>Список пользователей</h2>
        <button @click="fetchUsers">Обновить список</button>
        <ul>
          <li v-for="user in users" :key="user.id">
            <strong>Имя:</strong> {{ user.username }} <br />
            <strong>Источник данных:</strong> {{ user.dataSource }} <br />
            <strong>Уровень доступа:</strong> {{ user.permissionLevel }}
          </li>
        </ul>
      </div>
    </div>
  </template>
  
  <script>
  import { ref } from 'vue';
  
  export default {
    setup() {
      const newUser = ref({
        username: '',
        password: '',
        dataSource: '',
        permissionLevel: '1', // По умолчанию уровень доступа "Чтение"
      });
  
      const successMessage = ref('');
      const users = ref([]);
      let isUsersLoaded = false; // Флаг для кэширования списка пользователей
  
      const host = process.env.VUE_APP_HOST || 'http://localhost';
      const port = process.env.VUE_APP_SERVER_PORT || '3010';
      const url = `${host}:${port}/users`;
  
      console.log('Sending request to:', url); // Логирование URL
  
      // Функция для создания пользователя
      const createUser = async () => {
        if (!newUser.value.username || !newUser.value.password || !newUser.value.dataSource) {
          successMessage.value = 'Заполните все поля';
          return;
        }
  
        try {
          const response = await fetch(url, {
            method: 'POST',
            headers: {
              'Content-Type': 'application/json',
            },
            body: JSON.stringify(newUser.value),
          });
  
          if (!response.ok) {
            throw new Error('Ошибка при создании пользователя');
          }
  
          const data = await response.json();
          successMessage.value = `Пользователь "${data.username}" успешно создан с уровнем доступа "${data.permissionLevel}" и источником данных "${data.dataSource}"`;
          newUser.value = { username: '', password: '', dataSource: '', permissionLevel: '1' }; // Очистка формы
          fetchUsers(); // Обновляем список пользователей
        } catch (error) {
          console.error('Ошибка:', error);
          successMessage.value = 'Ошибка при создании пользователя';
        }
      };
  
      // Функция для получения списка пользователей
      const fetchUsers = async () => {
        if (isUsersLoaded) return; // Если список уже загружен, не делаем запрос
  
        try {
          const response = await fetch(url);
          if (!response.ok) {
            throw new Error('Ошибка при получении списка пользователей');
          }
          users.value = await response.json();
          isUsersLoaded = true; // Устанавливаем флаг
        } catch (error) {
          console.error('Ошибка:', error);
          successMessage.value = 'Ошибка при загрузке списка пользователей';
        }
      };
  
      // Загружаем список пользователей при монтировании компонента
      fetchUsers();
  
      return {
        newUser,
        successMessage,
        users,
        createUser,
        fetchUsers,
      };
    },
  };
  </script>
  
  <style>
  .success-message {
    color: green;
    margin-top: 10px;
  }
  </style>