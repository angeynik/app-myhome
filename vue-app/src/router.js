import { createRouter, createWebHistory } from 'vue-router';
import App from '@/App.vue';
import Info from '@/components/MainInfo.vue';
import Settings from '@/components/MainSettings.vue';
import Alarms from '@/components/MainAlarms.vue';
import Scenarios from '@/components/MainScenarios.vue';
import Statistic from '@/components/MainStatistic.vue';


const routes = [
  { path: '/', component: App},
  { path: '/info', component: Info },
  { path: '/settings', name: 'Settings', component: Settings },
  { path: '/alarms', component: Alarms },
  { path: '/scenarios', component: Scenarios },
  { path: '/statistic', component: Statistic },
];

const router = createRouter({
  history: createWebHistory(),
  routes,
});

export default router;
