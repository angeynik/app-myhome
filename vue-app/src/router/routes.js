import { createRouter, createWebHistory } from 'vue-router';
import Dashboard from '../DashBoard.vue';
import SmartHome from '../SmartHome.vue';
import ManufactAutomatation from '../ManufactAutomatation.vue';
import IntroduceHome from '../IntroduceHome.vue';

const routes = [
{
    path: '/',
    name: 'Intro',
    component: IntroduceHome, 
  },
  {
    path: '/dash-board',
    name: 'DashBoard',
    component: Dashboard,
  },
  {
    path: '/smart-home',
    name: 'SmartHome',
    component: SmartHome,
  },
  {
    path: '/manufact-automatation',
    name: 'ManufactAutomatation',
    component: ManufactAutomatation,
  },
];

const router = createRouter({
  history: createWebHistory(),
  routes,
});

export default router;