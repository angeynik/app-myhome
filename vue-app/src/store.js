import { reactive } from 'vue';

const state = reactive({
  isAuthenticated: false,
  user: null,
});

function login(user) {
  state.isAuthenticated = true;
  state.user = user;
}

function logout() {
  state.isAuthenticated = false;
  state.user = null;
}

export default {
  state,
  login,
  logout,
};
