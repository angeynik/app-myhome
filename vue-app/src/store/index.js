import { createStore } from 'vuex';
import auth from './modules/auth';
import websocket from './modules/websocket';
import log from './modules/log';
import sortParams from './modules/sortParams';
import config from './modules/config';

const store = createStore({
  modules: {
    auth,
    websocket,
    log,
    sortParams,
    config
  },
  actions: {
    toLowerCase(_, str) {
      return str.toLowerCase();
    },
  },
  getters: {
    isAuthenticated: (state) => !!state.auth.token,
    authStatus: (state) => state.auth.status,
    user: (state) => state.auth.user.username || '',
    level: (state) => state.auth.level || 0,
    dID: (state) => state.auth.dID || null,

  }
});

export default store;
