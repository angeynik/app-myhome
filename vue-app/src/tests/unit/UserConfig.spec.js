import { mount } from '@vue/test-utils';
import UserConfig from '@/components/UserConfig.vue';
import { createStore } from 'vuex';
import router from '@/router/routes'; // Импортируем ваш файл routes.js
import store from '@/store'; // Импортируем хранилище Vuex

describe('UserConfig.vue', () => {
  let testStore;

  beforeEach(() => {
    // Создаем мок store перед каждым тестом
    testStore = createStore({
      state: {},
      getters: {
        level: () => 2, // По умолчанию userLevel = 2
        isAuthenticated: () => true, // Пользователь авторизован
      },
      actions: {
        sendWSMessage: jest.fn((_, payload) => {
          if (payload.request === 'fetchUsers') {
            return Promise.resolve({ users: [{ id: 1, username: 'testUser' }] });
          }
          if (payload.request === 'fetchDataSources') {
            return Promise.resolve({ sources: [{ id: 1, did: 'testDataSource' }] });
          }
          if (payload.request === 'createUser') {
            return Promise.resolve();
          }
          if (payload.request === 'linkUserToDataSource') {
            return Promise.resolve();
          }
          return Promise.reject(new Error('Unknown request'));
        }),
        toLowerCase: jest.fn((_, value) => value.toLowerCase()),
      },
    });
  });

  it('matches the snapshot', async () => {
    const wrapper = mount(UserConfig, {
      global: {
        plugins: [testStore, router],
      },
    });
    await router.isReady(); // Ожидаем загрузки роутера
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('displays Profile link when userLevel >= 2', async () => {
    const wrapper = mount(UserConfig, {
      global: {
        plugins: [testStore, router],
      },
    });
    await router.isReady();
    expect(wrapper.find('a[href="/profile"]').exists()).toBe(true);
  });

  it('does not display Users link when userLevel < 3', async () => {
    const wrapper = mount(UserConfig, {
      global: {
        plugins: [testStore, router],
      },
    });
    await router.isReady();
    expect(wrapper.find('a[href="/users"]').exists()).toBe(false);
  });

  it('shows error message when username is too short', async () => {
    const wrapper = mount(UserConfig, {
      global: {
        plugins: [testStore, router],
      },
    });
    await router.isReady();

    await wrapper.find('#username').setValue('ab');
    await wrapper.find('form').trigger('submit.prevent');

    expect(wrapper.find('.error').text()).toContain('Имя пользователя должно быть от 3 до 16 символов.');
  });

  it('shows error message when password does not meet requirements', async () => {
    const wrapper = mount(UserConfig, {
      global: {
        plugins: [testStore, router],
      },
    });
    await router.isReady();

    await wrapper.find('#password').setValue('weakpassword');
    await wrapper.find('form').trigger('submit.prevent');

    expect(wrapper.find('.error').text()).toContain('Пароль должен быть от 10 до 32 символов, содержать минимум одну заглавную букву, цифру и спецсимвол.');
  });

  it('displays success message when user is created successfully', async () => {
    const wrapper = mount(UserConfig, {
      global: {
        plugins: [testStore, router],
      },
    });
    await router.isReady();

    await wrapper.find('#username').setValue('validUsername');
    await wrapper.find('#password').setValue('ValidPassword1!');
    await wrapper.find('#dataSource').setValue('valid_data_source');
    await wrapper.find('form').trigger('submit.prevent');

    expect(wrapper.find('.success-message').text()).toContain('Пользователь успешно создан');
  });

  it('displays success message when user is linked to data source successfully', async () => {
    const wrapper = mount(UserConfig, {
      global: {
        plugins: [testStore, router],
      },
    });
    await router.isReady();

    await wrapper.find('#existingUser').setValue('1');
    await wrapper.find('#existingDataSource').setValue('1');
    await wrapper.find('form').trigger('submit.prevent');

    expect(wrapper.find('.success-message').text()).toContain('Пользователь успешно связан с источником данных');
  });

  it('loads users and data sources on mount', async () => {
    const wrapper = mount(UserConfig, {
      global: {
        plugins: [testStore, router],
      },
    });
    await router.isReady();

    await wrapper.vm.fetchUsers();
    await wrapper.vm.fetchDataSources();

    expect(wrapper.vm.usersDB.length).toBeGreaterThan(0);
    expect(wrapper.vm.dataSources.length).toBeGreaterThan(0);
  });

  it('updates user list when refresh button is clicked', async () => {
    const wrapper = mount(UserConfig, {
      global: {
        plugins: [testStore, router],
      },
    });
    await router.isReady();

    await wrapper.find('button').trigger('click');
    expect(wrapper.vm.usersDB.length).toBeGreaterThan(0);
  });

  it('displays error message when user creation fails', async () => {
    testStore.dispatch = jest.fn(() => Promise.reject(new Error('Ошибка при создании пользователя')));
    const wrapper = mount(UserConfig, {
      global: {
        plugins: [testStore, router],
      },
    });
    await router.isReady();

    await wrapper.find('#username').setValue('invalidUsername');
    await wrapper.find('#password').setValue('invalidPassword');
    await wrapper.find('#dataSource').setValue('invalid_data_source');
    await wrapper.find('form').trigger('submit.prevent');

    expect(wrapper.find('.success-message').text()).toContain('Ошибка при создании пользователя');
  });

  it('displays error message when linking user to data source fails', async () => {
    testStore.dispatch = jest.fn(() => Promise.reject(new Error('Ошибка при связывании пользователя с источником данных')));
    const wrapper = mount(UserConfig, {
      global: {
        plugins: [testStore, router],
      },
    });
    await router.isReady();

    await wrapper.find('#existingUser').setValue('invalidUserId');
    await wrapper.find('#existingDataSource').setValue('invalidDataSourceId');
    await wrapper.find('form').trigger('submit.prevent');

    expect(wrapper.find('.success-message').text()).toContain('Ошибка при связывании пользователя с источником данных');
  });
});