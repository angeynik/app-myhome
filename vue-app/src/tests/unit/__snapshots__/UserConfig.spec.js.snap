// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserConfig.vue matches the snapshot 1`] = `
"<div>
  <header class=\\"header\\">
    <nav>
      <router-link to=\\"/login\\">Login</router-link> | <router-link to=\\"/dashboard\\">Dashboard</router-link> | <router-link to=\\"/smart-home\\">Smart Home</router-link> | <router-link to=\\"/manufact-automatation\\">Automation</router-link> | <router-link to=\\"/profile\\">Profile</router-link> |
      <!--v-if-->
    </nav>
    <p style=\\"width: 100%; height: 1px;\\"></p>
  </header><!-- Форма для создания пользователя -->
  <div style=\\"padding: 0px 10px 0px 10px;\\">
    <h2>Создать нового пользователя</h2>
    <form>
      <div><label for=\\"username\\">Имя пользователя:</label><input type=\\"text\\" id=\\"username\\" required=\\"\\" autocomplete=\\"username\\">
        <!--v-if-->
      </div>
      <div><label for=\\"password\\">Пароль:</label><input type=\\"password\\" id=\\"password\\" required=\\"\\" autocomplete=\\"password\\">
        <!--v-if-->
      </div>
      <div><label for=\\"dataSource\\">Источник данных:</label><input type=\\"text\\" id=\\"dataSource\\" required=\\"\\" autocomplete=\\"dataSource\\">
        <!--v-if-->
      </div>
      <div><label for=\\"permissionLevel\\">Уровень доступа:</label><select id=\\"permissionLevel\\" required=\\"\\" autocomplete=\\"permissionLevel\\">
          <option value=\\"1\\">Пользователь</option>
          <option value=\\"2\\">Оператор</option>
          <option value=\\"3\\">Администратор</option>
        </select></div><button type=\\"submit\\">Создать пользователя</button>
    </form>
  </div><!-- Форма для связывания пользователя с источником данных -->
  <div style=\\"padding: 0px 10px 0px 10px; margin-top: 20px;\\">
    <h2>Связать пользователя с источником данных</h2>
    <form>
      <div><label for=\\"existingUser\\">Пользователь:</label><select id=\\"existingUser\\" required=\\"\\"></select><!-- Лог для проверки данных -->
        <!-- <p>Список пользователей: {{ users }}</p> -->
      </div>
      <div><label for=\\"existingDataSource\\">Источник данных:</label><select id=\\"existingDataSource\\" required=\\"\\"></select><!-- Лог для проверки данных -->
        <!-- <p>Список источников данных: {{ dataSources }}</p> -->
      </div>
      <div style=\\"display: flex; justify-content: space-between;\\"><button>Обновить список</button><button type=\\"submit\\">Связать</button></div>
    </form>
  </div><!-- Сообщение об успешном создании пользователя -->
  <!--v-if-->
</div>"
`;
